auth/src/VSCodeAzureSubscriptionProvider.ts

```diff
@@ -0,0 +1,241 @@
+        const { client, credential, authentication } = await this.getSubscriptionClient(tenantId);
+        const environment = getConfiguredAzureEnv();
+
+        const subscriptions: AzureSubscription[] = [];
+
+        for await (const subscription of client.subscriptions.list()) {
+            subscriptions.push({
+                authentication: authentication,
+                environment: environment,
+                credential: credential,
+                isCustomCloud: environment.isCustomCloud,
+                /* eslint-disable @typescript-eslint/no-non-null-assertion */
+                name: subscription.displayName!,
+                subscriptionId: subscription.subscriptionId!,
+                /* eslint-enable @typescript-eslint/no-non-null-assertion */
+                tenantId: tenantId,
+            });
+        }
+
+        return subscriptions;
+    }
+
+    /**
+     * Gets a fully-configured subscription client for a given tenant ID
+     *
+     * @param tenantId The tenant ID to get a client for
+     *
+     * @returns A client, the credential used by the client, and the authentication function
+     */
+    private async getSubscriptionClient(tenantId: 'organizations' | string): Promise<{ client: SubscriptionClient, credential: TokenCredential, authentication: AzureAuthentication }> {
+        const armSubs = await import('@azure/arm-subscriptions');
+
+        // This gets filled in when the client calls `getToken`, and then it can be returned in the `authentication` property of `AzureSubscription`
+        let session: vscode.AuthenticationSession | undefined;
+
+        const credential: TokenCredential = {
+            getToken: async (scopes) => {
+                // TODO: change to `getSessions` when that API is available
+                session = await vscode.authentication.getSession(getConfiguredAuthProviderId(), this.getScopes(scopes, tenantId), { createIfNone: false, silent: true });
+                if (!session) {
+                    throw new NotSignedInError();
+                }
+
+                return {
+                    token: session.accessToken,
+                    expiresOnTimestamp: 0
+                };
+            }
+        }
+
+        return {
+            client: new armSubs.SubscriptionClient(credential),
+            credential: credential,
+            authentication: {
+                getSession: () => session // Rewrapped to make TS not confused about the weird initialization pattern
+            }
+        };
+    }
+
+    /**
+     * Gets a normalized list of scopes
+     *
+     * @param scopes An input scope string, list, or undefined
+     * @param tenantId (Optional) The tenant ID, will be added to the scopes
+     *
+     * @returns A list of scopes, with the default scope and (optionally) the tenant scope added
+     */
+    private getScopes(scopes: string | string[] | undefined, tenantId?: string): string[] {

```

Comment:
Tenant id will either be `'organizations'` or an actual tenant id. Never undefined.

```suggestion
    private getScopes(scopes: string | string[] | undefined, tenantId: string): string[] {
```

or 

```suggestion
    private getScopes(scopes: string | string[] | undefined, tenantId: 'organizations' | string): string[] {
```

If we go with this one I suggest extracting `'organizations'` as a new type.